### PR TITLE
DASHBUILDE-91: Filter UI Controls

### PR DESCRIPTION
**Pull request by dgutierr**
Includes a new displayer component "Selector" with three different subtypes: dropdown, label set & slider.  

This first iteration cover most of the use cases regarding the filtering of dashboards.

@romartin @jhrcek Can you review please?

PS: A video preview showing progress => https://youtu.be/6PZoHo7sMaI

### CODE REVIEW

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.dashbuilder.dataset.ExpenseReportsData.*;
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+public class SelectorSliderDisplayerTest extends AbstractDisplayerTest {
```
Comment (jhrcek) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java]: One more issue with a slider that I identified:
When upper and lower slider values are equal, it seems to be filtering data from some larger range as in the following screenshot: I guess that should be reported as well ?
screenshot from 2017-03-08 10-07-41

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java [hunk 2]
```diff
@@ -34,34 +33,36 @@ public AggregateFunctionType getType() {
...
-            if (n == null) continue;
-            if (max == null || n.doubleValue() > max.doubleValue()) max = n;
+    public Object aggregate(List values) {
+        if (values == null || values.isEmpty()) {
+            return null;
         }
-        if (max == null) return 0;
-        return round(max.doubleValue(), precission);
+
+        List sorted = (List) values.stream().sorted().collect(Collectors.toList());
```
Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: As mentioned above: Please filter out nulls and provide some unit tests to cover those cases.

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MinFunction.java [hunk 3]
```diff
@@ -34,36 +34,36 @@ public AggregateFunctionType getType() {
...
-            if (min == null || n.doubleValue() < min.doubleValue()) min = n;
+    public Object aggregate(List values) {
+        if (values == null || values.isEmpty()) {
+            return null;
         }
-        // Adjust to the specified precision.
-        if (min == null) return 0;
-        return round(min.doubleValue(), precission);
+
+        List sorted = (List) values.stream().sorted().collect(Collectors.toList());
```
Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MinFunction.java]: The same here.

File: dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java [hunk 4]
```diff
@@ -108,6 +109,7 @@ public void init(DataSetMetadata metadata,
         this.groupFunction = groupFunction;
         this.targetType = targetType;
         this.metadata = metadata;
+        this.functionsEnabled = functionsEnabled;
 
         columnDetailsEditor.init(metadata, this.groupFunction);
```
Comment (jhrcek) [inline dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java]: IntelliJ static analysis discovered an initialization issue in [related code](https://github.com/dashbuilder/dashbuilder/blob/master/dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnDetailsEditorView.java#L46)
Instead of 
`this.presenter = presenter`
is should be 
`this.presenter = columnDetailsEditor`, otherwise the presenter remains uninitialized

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/main/java/org/dashbuilder/renderer/client/selector/SelectorLabelSetDisplayer.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+import org.dashbuilder.dataset.sort.SortOrder;
+import org.dashbuilder.displayer.ColumnSettings;
+import org.dashbuilder.displayer.DisplayerAttributeDef;
+import org.dashbuilder.displayer.DisplayerAttributeGroupDef;
+import org.dashbuilder.displayer.DisplayerConstraints;
+import org.dashbuilder.displayer.client.AbstractErraiDisplayer;
+import org.jboss.errai.ioc.client.container.SyncBeanManager;
+
+@Dependent
+public class SelectorLabelSetDisplayer extends AbstractErraiDisplayer<SelectorLabelSetDisplayer.View> {
```
Comment (jhrcek) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/main/java/org/dashbuilder/renderer/client/selector/SelectorLabelSetDisplayer.java]: There is one potential UX issue with this new SelectorLabel displayer: it would be nice to have something similar to "X" button which we have for Multiple SelectorDropdown displayer which would clean all the selected labels and would bring the selector to its initial state (no labels selected). Otherwise user has to "unclick" each of the selected labels individually if he wants reset to initial state. Could you implement that, please?

File: dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/resources/images/DisplayerImagesResources.java [hunk 6]
```diff
@@ -139,6 +139,24 @@
     @Source("metric_plaintext_unselected.png")
     ImageResource METRIC_METRIC_PLAIN_TEXT_unselected();
 
+    @Source("selector_dropdown_selected.png")
```
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/resources/images/DisplayerImagesResources.java]: I'll review them.

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java (hunk 2 above)
Comment (dgutierr) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: Sure. I will.

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java (hunk 1 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java]: I will review that as well.

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MinFunction.java (hunk 3 above)
Comment (dgutierr) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MinFunction.java]: Ok

File: dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java (hunk 4 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java]: ok

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/main/java/org/dashbuilder/renderer/client/selector/SelectorLabelSetDisplayer.java (hunk 5 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/main/java/org/dashbuilder/renderer/client/selector/SelectorLabelSetDisplayer.java]: Adding that to the TODO list. I'd like to see what UXD thinks about. 

Thanks for the thorough review @jhrcek :-)

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java (hunk 1 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java]: Done

File: dashbuilder-shared/dashbuilder-dataset-shared/src/test/java/org/dashbuilder/dataset/engine/function/AggregateFunctionsTest.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+public class AggregateFunctionsTest {
+
+    DateFormat DATE_FORMAT = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss");
+    List<String> STRING_LIST = Arrays.asList("a", "b", "c");
+    List<String> STRING_LIST_WITH_NULL = Arrays.asList("a", "b", "c", null);
+    List<Date> DATE_LIST = new ArrayList<>();
+    List<Date> DATE_LIST_WITH_NULL = new ArrayList<>();
+    List<Double> NUMERIC_LIST = Arrays.asList(1d, 2d, 3d);
+    List<Double> NUMERIC_LIST_WITH_NULL = Arrays.asList(1d, 2d, 3d, null);
+    Date date1;
```
Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/test/java/org/dashbuilder/dataset/engine/function/AggregateFunctionsTest.java]: Nice coverage! But there is one edge case which wasn't taken into consideration: what if ALL values in the list are null? In such case the aggregation functions call `sorted.get(0)` which would throw `IndexOutOfBoundException`.

I think this should be covered in aggregation functions implementation as well (e.g. imagine someone adds new values to a DB column when doing migration and sets all cells to null). Could you please cover that case as well?

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java [hunk 8]
```diff
@@ -38,7 +38,10 @@ public Object aggregate(List values) {
             return null;
         }
 
-        List sorted = (List) values.stream().sorted().collect(Collectors.toList());
+        List sorted = (List) values.stream()
+                .filter(i -> i != null)
+                .sorted().collect(Collectors.toList());
```
Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: Is it really necessary to sort the entire list (O(n*log(n)) complexity just to get the max value?
Can't for example [max](http://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#max-java.util.Comparator-) to get the same result in linear time? The same comment applies to MinFunction.

File: dashbuilder-shared/dashbuilder-dataset-shared/src/test/java/org/dashbuilder/dataset/engine/function/AggregateFunctionsTest.java (hunk 7 above)
Comment (dgutierr) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/test/java/org/dashbuilder/dataset/engine/function/AggregateFunctionsTest.java]: Done.

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java (hunk 8 above)
Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: The difference still matters. Whereas max touches each element only once, sorting in general has to touch each element multiple times. I encourage you to explore the pefrormance impact of that for yourself using this [microbenchmark](https://gist.github.com/jhrcek/16308625c2e8a8a0b0267d367ab6d31d) for different values of N.

Comment (jhrcek) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: Edit: i might be wrong about what's causing the performance difference (it might be collecting the sorted data `toList()`), but the difference is noticeable with larger data sets (100k+ items) especially when multiple aggregation functions for different columns are used.

File: dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/resources/images/DisplayerImagesResources.java (hunk 6 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/resources/images/DisplayerImagesResources.java]: Done

File: dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java (hunk 1 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-default/src/test/java/org/dashbuilder/renderer/client/selector/SelectorSliderDisplayerTest.java]: Done

File: dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java (hunk 4 above)
Comment (dgutierr) [inline dashbuilder-client/dashbuilder-displayer-client/src/main/java/org/dashbuilder/displayer/client/widgets/group/ColumnFunctionEditor.java]: Done

File: dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java (hunk 8 above)
Comment (dgutierr) [inline dashbuilder-shared/dashbuilder-dataset-shared/src/main/java/org/dashbuilder/dataset/engine/function/MaxFunction.java]: I've changed the current impl to avoid the sorting of the entire collection.

**Comment by dgutierr** [comment]
@jhrcek Added the changes you suggested.

jenkins retest this

**Comment by jhrcek** [comment]
@dgutierr I did the 2nd round of review and all the issues I found have been fixed except for one:

I can still reproduce the problem with slider displayer - when lower and upper limits are equal I'm getting values from some wider interval. Here is the video illustrating the issue
[sliderIssue.zip](https://github.com/dashbuilder/dashbuilder/files/857305/sliderIssue.zip) with data from csv dataset as added to the [previous jira](https://issues.jboss.org/browse/DASHBUILDE-209)

**Comment by dgutierr** [comment]
@jhrcek The latest issue you mention is also fixed.

I've also applied some fine tuning, specially in the dropdown & slider selectors as recommended by UXD team =>https://docs.google.com/document/d/1YawDcudXkCEq3uVh1h-J-xd_uyd8BROqSlUaWhrh9T4/edit#    Notice, the document introduces some advanced features which will be implemented in future versions.

A video showing the latest changes => https://issues.jboss.org/secure/attachment/12418440/dashbuilder_selectors_finetune.mp4

@romartin In my opinion, the PR is ready to merge.

**Comment by jhrcek** [comment]
I tested it locally and all the issues I found have been fixed. +1 to merge

**Comment by romartin** [comment]
Closed: 5666fa9af4a3c5bd8df9026053a610641797a1c5
